### PR TITLE
 Deleting Students from the Subject as a Teacher #72

### PR DESCRIPTION
**Pull request by fedyafedko**

### CODE REVIEW

File: StudyHub/StudyHub.BLL/Services/SubjectService.cs [hunk 1]
```diff
@@ -121,4 +95,66 @@ public async Task<SubjectDTO> UpdateSubjectAsync(Guid userId, Guid subjectId, Up
...
         return _mapper.Map<SubjectDTO>(entity);
     }
+
+    public async Task<List<StudentDTO>> AddStudentsToSubjectAsync(
+        Guid subjectId,
+        Guid teacherId,
+        StudentsToSubjectRequest request)
+    {
+        var subject = await _subjectRepository.FirstOrDefaultAsync(s => s.Id == subjectId)
+            ?? throw new NotFoundException("Subject not found with the specified Id");
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/SubjectService.cs:105]: Probably better to add current Id in Exception which is not found

File: StudyHub/StudyHub.BLL/Services/SubjectService.cs [hunk 2]
```diff
@@ -121,4 +95,66 @@ public async Task<SubjectDTO> UpdateSubjectAsync(Guid userId, Guid subjectId, Up
...
+
+        if (subject.TeacherId != teacherId)
+            throw new RestrictedAccessException("You are not the owner and do not have permission to perform this action.");
+
+        foreach (var email in request.Emails)
+        {
+            var user = await _userManager.FindByEmailAsync(email)
+                ?? throw new NotFoundException($"Student not found with the specified email: {email}");
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/SubjectService.cs:113]: Let's image, we have 3 emails which we wanted to add, first pass correctly, second will no found in user manager, what happen to third one? And also, we partly have succesfull adding, but we will have an Exception. Think, needed some speculate here

File: StudyHub/StudyHub.BLL/Services/SubjectService.cs [hunk 3]
```diff
@@ -121,4 +95,66 @@ public async Task<SubjectDTO> UpdateSubjectAsync(Guid userId, Guid subjectId, Up
...
+                .FirstOrDefaultAsync(s => s.Id == subjectId)
+            ?? throw new NotFoundException("Subject not found with the specified Id");
+
+        if (subject.TeacherId != teacherId)
+            throw new RestrictedAccessException("You are not the owner and do not have permission to perform this action.");
+
+        foreach (var email in request.Emails)
+        {
+            var user = await _userManager.FindByEmailAsync(email)
+                ?? throw new NotFoundException($"Student not found with the specified email: {email}");
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/SubjectService.cs:141]: same here

File: StudyHub/StudyHub.BLL/Services/Interfaces/ISubjectService.cs [hunk 4]
```diff
@@ -17,5 +16,7 @@ public interface ISubjectService
 
     Task<List<SubjectDTO>> GetSubjectsForUserAsync(Guid userId);
 
-    Task<List<StudentDTO>> AddStudentsToSubjectAsync(Guid subjectId, Guid teacherId, AddStudentsToSubjectRequest request);
+    Task<List<StudentDTO>> AddStudentsToSubjectAsync(Guid subjectId, Guid teacherId, StudentsToSubjectRequest request);
+    Task<bool> DeleteStudentsSubjectAsync(Guid subjectId, Guid teacherId, StudentsToSubjectRequest request);
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/Interfaces/ISubjectService.cs:20]: DeleteStudentsFromSubjectAsync*

File: StudyHub/StudyHub.BLL/Services/SubjectService.cs [hunk 5]
```diff
@@ -121,4 +95,66 @@ public async Task<SubjectDTO> UpdateSubjectAsync(Guid userId, Guid subjectId, Up
...
+
+            user.Subjects ??= new List<Subject>();
```
Comment (Podgore) [inline StudyHub/StudyHub.BLL/Services/SubjectService.cs:115]: not really get why you create SubjectList. I got, that you not inlude navigation prop Subject(idk if we can do it with userManager), but I wanted to ask, if it working correctly in that case if we already got some subjects in user. I think, that it delete(probably) all current subject and put only one that you got here. Double check it please

**Review by Podgore — CHANGES_REQUESTED** [review]

File: StudyHub/StudyHub.BLL/Services/SubjectService.cs (hunk 5 above)
Comment (fedyafedko) [reply StudyHub/StudyHub.BLL/Services/SubjectService.cs:115]: it not delete all current subject because i use ??=

**Review by Podgore — DISMISSED** [review]

**Review by aquaprogit — CHANGES_REQUESTED** [review]

File: StudyHub/StudyHub/Controllers/SubjectController.cs [hunk 6]
```diff
@@ -83,7 +83,7 @@ public async Task<IActionResult> AddStudentsToSubject(Guid subjectId, StudentsTo
         return Ok(result);
     }
 
-    [HttpDelete("[action]/{subjectId}")]
+    [HttpDelete("students/{subjectId}")]
```
Comment (aquaprogit) [inline StudyHub/StudyHub/Controllers/SubjectController.cs:86]: subject/{id}/students looks better, same for POST

**Review by aquaprogit — CHANGES_REQUESTED** [review]

**Review by aquaprogit — APPROVED** [review]

**Pull request merged**